### ISSUE TITLE
10.2.3 -> 11.1.3 increase CPU usage (from ~40% to 90-150% for grafana)

### ISSUE THREAD

**Issue by bob454522** [grafana/grafana#81401]
### What happened?

OS: Debian 8.11
Grafana version: 10.3.1 (was 10.2.3 prior)

Ive noticed a near doubling of cpu usage for the grafana process (~ 40% to ~80% constant), with the only change being i updated from 10.2.3 to 10.3.1.  (see graph below please).  

I do have 3x larger dashboards being displayed 24/7 , each with ~45sec refresh interval set (this is the same case as was on 10.2.3)

I do keep deep stats on my VMs (thanks to grafana!)  , and it does seem that along with the increase in CPU, there has been a large increase in Writes by grafana (kB-wr/sec).

(log level has always been set to level warn,  I did try setting it to debug temporarily and did not see anything in those debug logs relevant, imo)

I have searched around and found that a sTrace It's frequently requested for debugging,  so that info is below, along with some charts.

Please let me know what additional info i can provide to help with this.

thanks
(LOVE GRAFANA!)

EDIT:i should add these relevant entries from my grafana.ini: 
```
[metrics]
enabled = disable

[log]
mode = file
level = warn
;level = debug
``` 

![image](https://github.com/grafana/grafana/assets/60830628/e1c81468-eaaa-4370-8eaa-07f1c0a5e805)

![image](https://github.com/grafana/grafana/assets/60830628/e38ef177-9ff8-430d-9bac-0407ab189ea7)

strace output (not sure if i ran this correctly,  but 25773 is grafana's PID  , and i let it run for about 20sec before CTRL + C):

```
root@XX:~# lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.11 (jessie)
Release:        8.11
Codename:       jessie

root@XX:~# strace -cfp 25773
Process 25773 attached with 28 threads
Process 37724 attached
Process 37809 attached

... ( SEVERAL of these i have clipped ) ...

Process 39626 detached
Process 39627 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 57.28   92.894565        6271     14813         2 nanosleep
 39.74   64.445202         122    529209     10954 futex
  1.32    2.146384           4    485015           pread
  0.82    1.322054         421      3144           epoll_pwait
  0.63    1.026297         369      2781      2056 stat
  0.05    0.088023         125       704           lstat
  0.04    0.072108           1     55147           sched_yield
  0.04    0.058263          37      1565           fstat
  0.03    0.044295          58       758           getpid
  0.03    0.041635          40      1052       423 read
  0.00    0.007234           8       912           setsockopt
  0.00    0.004433           1      3206           munmap
  0.00    0.004000           6       719         1 write
  0.00    0.004000           7       557        16 rt_sigreturn
  0.00    0.002611           1      4149           mmap
  0.00    0.002533           9       295           close
  0.00    0.000210           1       272           pwrite
  0.00    0.000141           0      1099           madvise
  0.00    0.000000           0       164           open
  0.00    0.000000           0        50           mprotect
  0.00    0.000000           0        23           brk
  0.00    0.000000           0       300           rt_sigprocmask
  0.00    0.000000           0         2           writev
  0.00    0.000000           0         4           mremap
  0.00    0.000000           0        14           socket
  0.00    0.000000           0        14        14 connect
  0.00    0.000000           0       228           getsockname
  0.00    0.000000           0        13           getpeername
  0.00    0.000000           0        13           getsockopt
  0.00    0.000000           0        50           clone
  0.00    0.000000           0      1937           fcntl
  0.00    0.000000           0        78           fsync
  0.00    0.000000           0        26           unlink
  0.00    0.000000           0        26           fchmod
  0.00    0.000000           0        26           geteuid
  0.00    0.000000           0       100           sigaltstack
  0.00    0.000000           0        50           gettid
  0.00    0.000000           0         1           restart_syscall
  0.00    0.000000           0       430         1 epoll_ctl
  0.00    0.000000           0       556           tgkill
  0.00    0.000000           0         1           openat
  0.00    0.000000           0       359       143 accept4
------ ----------- ----------- --------- --------- ----------------
100.00  162.163988

```

### What did you expect to happen?

similar cpu usage on 10.3.1  as was the case on 10.2.3

### Did this work before?

did not see this issue on 10.2.3

### How do we reproduce it?

See post above (i just updated from 10.2.3 -> 10.3.1

### Environment (with versions)?

Grafana: 10.3.1
OS: Debian 8.11
Browser:

### Grafana platform?

A downloaded binary

**Comment by usmangt** [comment]
Hi @bob454522,

Thanks for opening this issue.

Could you please change the default log level to debug mode? 

To do this go to the server where grafana is installed and edit the configuration file (`/etc/grafana/grafana.ini`)

There you will this section;

```
#################################### Logging ##########################
[log]
# Either "console", "file", "syslog". Default is console and  file
# Use space to separate multiple modes, e.g. "console file"
;mode = console file

# Either "debug", "info", "warn", "error", "critical", default is "info"
;level = info
```
Set the parameter level to debug and remove the colon in the beginning i.e.

`level = debug`

Then restart the grafana service and resent the logs once the issue occurs again.

**Comment by bob454522** [comment]
> Could you please change the default log level to debug mode?

> Then restart the grafana service and resent the logs once the issue occurs again.

Thanks for the reply,  below are the requested logs at debug level,  one note-  You said to send these logs "_once the issue occurs again_", to be clear the issue is constant, the CPU usage is constantly elevated  (versus before on 10.2.3 it was lower, as seen in graphs i posted above).

note- this grafana instance is only accessible on my private/local network (ie is not expose to or accessible from the public internet)

(i did redact a few thing in the logs,  i replaced those with XXX ) 
please let me know if i can provide anything further,
thanks


[grafana.log](https://github.com/grafana/grafana/files/14103683/grafana.log)

**Comment by usmangt** [comment]
Thanks for the logs :+1: 

I will forward this to our Engineers so that they can have a look and might be able to tell more details about it.

**Comment by bob454522** [comment]
> Thanks for the logs 👍
> 
> I will forward this to our Engineers so that they can have a look and might be able to tell more details about it.

thanks,  im re-reading and in my first post the axis is cut off so it's not clear the CPU increase from version 10.2.3 to version 10.3.1 ,  so here is that same graph with the axis more clear:
LOVE GRAFANA!

![image](https://github.com/grafana/grafana/assets/60830628/66d532fd-3c46-4af5-902a-437e2cd9084f)

**Comment by bob454522** [comment]
I wanted to add:  Earlier today i tried installing the 10.4.0-65999 (nightly),  to see if there was any difference in the cpu usage- and the **issue does still persist** (exact same elevated cpu usage,  no more or less on 10.3.1 vs the 10.4.0 nightly).  So it seems its was something that changed from 10.2.3  to 10.3.1  (or maybe 10.3.0 but i can stay that for sure).

Im on enterprise, doubt this matters but fyi.
(I do not have a license).
thanks
LOVE GRAFANA!

**Comment by papagian** [comment]
@bob454522 thank you for reporting this!
could you share which SQL database do you use?

**Comment by bob454522** [comment]
> @bob454522 thank you for reporting this! could you share which SQL database do you use?
> 
> also if it's possible, could you [enable profiling](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/configure-tracing/#configure-tracing-to-troubleshoot-grafana) in your grafana and share the profiles?

>
will update with profiling when i get a chance,   thanks for guidance -
as for SQL db -  im using the default ,  so im pretty sure that is sqlite.    here is from my grafana.ini:

```
#################################### Database ############################
[database]
# You can configure the database connection by specifying type, host, name, user and password
# as separate properties or as on string using the url property.

# Either "mysql", "postgres" or "sqlite3", it's your choice
type = sqlite3
host = 127.0.0.1:3306
name = grafana
user = xxx
# If the password contains # or ; you have to wrap it with triple quotes. Ex """#password;"""
password = 
# Use either URL or the previous fields to configure the database
# Example: mysql://user:secret@host:port/database
url =

# Max idle conn setting default is 2
max_idle_conn = 2

# Max conn setting default is 0 (mean not set)
max_open_conn =

# Connection Max Lifetime default is 14400 (means 14400 seconds or 4 hours)
conn_max_lifetime = 14400

log_queries =

ssl_mode = disable

isolation_level =

ca_cert_path =
client_key_path =
client_cert_path =
server_cert_name =

path = grafana.db

cache_mode = private

wal = false

migrations, default is 0.
locking_attempt_timeout_sec = 0

query_retries = 0
transaction_retries = 5
instrument_queries = false
```
( i did remove some of the  # comment lines when i pasted above , to keep it shorter)

**Comment by bob454522** [comment]
versions 10.4.x ( 10.4.1 and .2)  and also 11-preview  all exhibit this same high cpu usage  (exact same as i describe above,  no more or no less cpu usage vs versions after 10.2.3).
![image](https://github.com/grafana/grafana/assets/60830628/9c8d259c-64ef-406d-a082-0da819bfdda6)

So im trying to grab the go profile as requested so i can share it here.

I installed go (```go -h``` works) ,  and am trying to follow these directions:
https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/configure-tracing/#turn-on-profiling

2x questions ive run into:
1- that first command listed in the docs should have a DASH,  right?  (just want to be sure)
./**grafana-server** -profile -profile-addr=0.0.0.0 -profile-port=8080
not:
./**grafana server** -profile -profile-addr=0.0.0.0 -profile-port=8080

#2-  im not clear on this command:
```go tool pprof binary profile```

once i have restarted grafana with these flags ,  ie ( ```./**grafana-server** -profile -profile-addr=0.0.0.0 -profile-port=8080```   ,  do i then run:
```go tool grafana-server profile```

(i dont have experience with GO ,and the docs are not clear on this),  when i try the above i get:

![image](https://github.com/grafana/grafana/assets/60830628/cb453611-717d-4809-90ac-27e7666b4e12)

thanks

**Comment by bob454522** [comment]
can anyone please help with the ```go tool grafana-server profile``` question i have (so that i can provide more details to this thread).

I do still see the same high cpu usage on 11.1.3.     my question is how do i get the go profiling data after running:
```./grafana-server -profile -profile-addr=0.0.0.0 -profile-port=8080 ```
(please see my reply above this with details of what i tried).
thanks

**Comment by bob454522** [comment]
I have a new very interesting data point in this issue;   _(bottom line/TLDR: - new os / new grafana install / same use-case = same high cpu usage for grafana )_

today i created a brand new debian12 install and did a new grafana install version 11.1.3 (**on different, still server grade, hardware** from the grafana server with cpu usage issue above).

I did copy the grafana.ini and grafana.db files over (so i could test with similar dashboards), but id left off all plugins (of which im only using 4x).

I then changed the IP of this new debian12/grafana machine to my main grafana server (so that my various Grafana Live streaming stats would get sent, and my dashboards would show up)-  **instantly the grafana cpu usage shot up, just like on my existing Debian 8 "main" grafana server**  (see image below pls)

so its not something unique to my install nor due to using the older OS (deb8 / old dependencies).

Any suggestions will be would be greatly appreciated (i have the trace log posted at the top of this thread) , in the meantime I'm going to try to get go cpu profiling working on this secondary Debian12 install against grafana. 

![image](https://github.com/user-attachments/assets/3fa2590c-c44c-432b-b0fe-c27514bd9841)

**Comment by bob454522** [comment]
ok @papagian here is the requested go cpu profile, hopefully this will help ID the issue.

[cpu_profile.zip](https://github.com/user-attachments/files/16606504/cpu_profile.zip)
[profile.pb.gz](https://github.com/user-attachments/files/16606505/profile.pb.gz)


im also going to Include the steps I took to get this profile so hopefully it'll help others, as the grafana docs on this are lacking, and it seems it's incredibly difficult to launch grafana directly from the command line (others have mentioned this, im sure there are reason though) so i temporarily modified the service file:

```
nano /usr/lib/systemd/system/grafana-server.service
(if this above is not the path to your service file,  run: 
systemctl cat grafana-server 
and it will show you the correct path)

add into the service file (right after the EnviromentFile=  YOU MAY HAVE TO ADJUST THE 5080 PORT,  8080 IS DEFAULT):

Environment="GF_DIAGNOSTICS_PROFILING_ENABLED=true"
Environment="GF_DIAGNOSTICS_PROFILING_ADDR=0.0.0.0"
Environment="GF_DIAGNOSTICS_PROFILING_PORT=5080"

sudo systemctl daemon-reload
sudo systemctl restart grafana-server
sudo systemctl status grafana-server

then restart the service and run this:
curl -o cpu_profile.prof http://localhost:5080/debug/pprof/profile?seconds=30

go tool pprof cpu_profile.prof

(or to see a web visualization ) 
go tool pprof -http="0.0.0.0:8085" cpu_profile.prof

```

where to add the env vars in the service file:
![image](https://github.com/user-attachments/assets/9434e624-a122-4d42-ba5e-674169319e0a)

how to know that pprof it enabled (you only want to run this t Temporarily while troubleshooting not permanently ):
![image](https://github.com/user-attachments/assets/a7ca4242-3a7f-4b35-b1b1-dbe054b48b0b)

**Comment by bob454522** [comment]
i realize this thread has gotten long and un-organized -  to try to sum up,  here is the relevant info currently:

below is the requested go cpu profile while grafana was showing 80-130% constant cpu usage (was recorded on from my original, main, deb8-jessie grafana11.1.3 server) , hopefully this will help ID the issue.

[cpu_profile.zip](https://github.com/user-attachments/files/16606504/cpu_profile.zip)
[profile.pb.gz](https://github.com/user-attachments/files/16606505/profile.pb.gz)

im not sure how to read these Visualizations (i'm not a coder specifically), but image below is of **what stands out to me (The visualization is of a 30sec capture via:  /debug/pprof/profile?seconds=30**   ) :

(the ```span_name HTTP POST:``` are a few metrics im pushing to the Grafana Live Streaming system (at about 20-100 messages per 1minute so not very fast):
thank you

![image](https://github.com/user-attachments/assets/b4f15c2c-c195-43fe-9a03-01582f0feb4e)

**Comment by bob454522** [comment]
note- from that cpu profile chart (from go cpu profile of grafana) - it looks like something related to grafana live streaming (maybe?)  - so i Tried switching the endpoints that are logging to those http post urls to using Grafana live via websockets,  (ie see docs here:

However the cpu usage of Grafana Server and the load average of this machine remained pretty much the same.  I can say as I was doing this I had to stop the various live streaming end points (ie the producers send data to grafana live) for about 20 to 30 minutes and during that time I saw about a 30% reduction in the grafana processes CPU usage  (i was expected alot more of a reduction however ,  as the usage was still very high even without grafana live having data sent to it).  

(im not a coder nor proficient in go).
thanks
